### DISCUSSION TITLE
How to do integration testing in `axum` ?

### DISCUSSION THREAD

**Discussion by roynrishingha** [tokio-rs/axum#1423] [Q&A] [answered]
# Integration Testing

How can we perform **integration test** on our `axum` app?

```
...
|── src
│   ├── lib.rs
│   ├── main.rs
|── tests
│   ├── integration_test_one.rs
│   ├── health_check.rs
...
```
How can we do this?

Note: If my question can be improved, please suggest me as well.

**Comment by davidpdrsn** [comment, marked as answer]
Using "integration tests" (i.e. tests stored in the `tests` directory rather than `src`) is tricky because it requires you expose things (via `pub`) from your crate. In most cases if you're building a binary that isn't necessary. This is mainly intended for libaries.

I recommend you write tests in directly in `src` like its done in this example https://github.com/tokio-rs/axum/blob/main/examples/testing/src/main.rs

**Reply by roynrishingha** [reply]
Hello @davidpdrsn
I'm curious to know how to do it. Isn't it possible to write tests in `test` directory?

Thanks

**Reply by davidpdrsn** [reply]
> Isn't it possible to write tests in test directory?

I feel like I answered that question already. Yes its possible, but intended for library developers.

Do this https://github.com/tokio-rs/axum/blob/main/examples/testing/src/main.rs instead.
